### ISSUE TITLE
How to convert a LazyTensor into a dense torch.Tensor before reduction?

### ISSUE THREAD

**Issue by srikanthmalla** [getkeops/keops#88]
Hi @jeanfeydy ,
I am new to this library. I am running your example of clustering (https://www.kernel-operations.io/keops/_auto_tutorials/kmeans/plot_kmeans_torch.html).

```python
def KMeans(x, K=10, Niter=10, verbose=True):
     ......
    for i in range(Niter):
        ........
        D_ij = ((x_i - c_j) ** 2).sum(-1)  # (Npoints, Nclusters) symbolic matrix (lazy tensor)
        cl = D_ij.argmin(dim=1).long().view(-1)  # Points -> Nearest cluster (pytorch tensor)
        ......
    return cl, c
```

Is there a way to access D_ij result as a pytorch tensor (before reduction with argmin or min) and return similar to cl ? Please let me know.

Thank you,
Srikanth

**Comment by srikanthmalla** [comment]
Hi @jeanfeydy ,
Thank you for the prompt reply. I am trying to re-implement deep cluster-v2 with keops, to use with my project:

deep cluster-v2 original implementation: https://github.com/facebookresearch/swav/blob/d4970c83791c8bd9928ec4efcc25d4fd788c405f/main_deepclusterv2.py#L343
paper: 
https://arxiv.org/pdf/1807.05520.pdf

1. I am currently computing the D_ij to dense tensor using dot product by multiplication with identity matrix. 
->ones = torch.eye(K).to(args.device)
->scores = (D_ij@ones).float() #this operation **takes lot of time like 8 sec**, otherwise the whole function takes 0.08 sec

2. Also ideally I want to change the distance metric to dot product (according to the paper). But I get empty clusters when using dot product distance, I have to assign non-empty cluster centroid with some perturbation to empty clusters (according to the paper, to overcome this issue).
if args.dist_metric=="euclidean":
        D_ij = ((x_i - c_j) ** 2).sum(-1)  # (Npoints, Nclusters) symbolic matrix of squared distances
if args.dist_metric=="cosine":
        D_ij = x_n@c.t() # dot product by mat-mul, x_n and c.t are normalized 

Any thoughts or suggestions will be very helpful.

Thank you,
Srikanth

**Comment by jeanfeydy** [comment]
Hi @srikanthmalla ,

You're welcome! Some answers to your questions:

1. In the link above, I only see a (surprisingly complex and heavy) implementation of the K-means algorithm. Could you please tell me at which line you need to compute the **full matrix of distances** between the centroids and the samples?

2. KeOps allows you to work with dot products easily: the key here is to use the **dot product** operator `(x_i | y_j)`. A typical implementation of the K-means algorithm with the cosine similarity "metric" reads:

```python
import torch
from matplotlib import pyplot as plt

from pykeops.torch import LazyTensor

def KMeans(x, K=10, Niter=10):
    N, D = x.shape  # Number of samples, dimension of the ambient space

    # K-means loop:
    # - x  is the point cloud,
    # - cl is the vector of class labels
    # - c  is the cloud of normalizes cluster centroids
    c = x[:K, :].clone()  # Simplistic random initialization

    # Normalize the centroids, since we work with the cosine similarity "metric":
    c = torch.nn.functional.normalize(c, dim=1, p=2)

    # Symbolic representation:
    x_i = LazyTensor(x[:, None, :])  # (Npoints, 1, D)
    c_j = LazyTensor(c[None, :, :])  # (1, Nclusters, D)

    for i in range(Niter):

        # E step: assign points to a cluster that maximizes correlation --------
        D_ij = (x_i | c_j)  # (Npoints, Nclusters) symbolic Gram matrix
        cl = D_ij.argmax(dim=1).long().view(-1)  # Points -> Nearest cluster
        
        # M step: update the centroids to the normalized cluster average: ------
        # Average centroid per cluster:
        c.zero_()
        c.scatter_add_(0, cl[:,None].repeat(1, D), x)

        # Normalize the centroids, in place:
        c[:] = torch.nn.functional.normalize(c, dim=1, p=2)

    return cl, c
```

You can use it as follows:
```python
N, D, K = 10000, 2, 50
x = .4 * torch.randn(N, D).cuda() + .3

cl, c = KMeans(x, K)
plt.figure(figsize=(8,8))
plt.scatter(x[:, 0].cpu(), x[:, 1].cpu(), c=cl.cpu(), s= 30000 / len(x), cmap="tab10")
plt.scatter(c[:, 0].cpu(), c[:, 1].cpu(), c='black', s=50, alpha=.8)
plt.axis([-1.2,1.2,-1.2,1.2]) ; plt.tight_layout() ; plt.show()
```

![kmeans_cosine](https://user-images.githubusercontent.com/20405678/91089944-96ced480-e654-11ea-9915-0e071fa8d672.png)

If needed, you can detect empty clusters using the bincount routine:
```python
# Number of points per cluster:
Ncl = torch.bincount(cl, minlength=K)
empty = (Ncl == 0)
```
Using standard PyTorch indexing, you will then be able to assign fresh values to the cluster-less centroids `c[empty,:]`, using your favorite heuristic.

What do you think?

Best regards,
Jean

**Comment by srikanthmalla** [comment]
Hi @jeanfeydy ,
Thanks so much for the detailed explanation.
1. I think this `D_ij = (x_i | c_j)` is what I'm looking for cosine similarly.

2. You are right the scores should be output of the network to penalize according to the paper. Then I will not be needing the D_ij dense array for penalization (which I initially thought of doing, which is wrong). But what I'm thinking is to order the top n closest points to a latent feature (with in the same cluster) for multimodal outputs with ranking (for a prediction task using regression) in the order of closest latent features during inference. Which I will call in a separate a function, but still the question of computing the dense array in a fast manner (like other keops operations) arises. Please let me know if that is possible or if something is not clear.

Thank you very much,
Srikanth

**Comment by jeanfeydy** [comment]
Good!

For your second point: the `.argKmin(K=...)` reduction (as well as `Kmin_argKmin`, etc.) is what you are looking for. It is equivalent to PyTorch's `.topk()` method, but handles KeOps LazyTensors and is usually much faster: its usage is explained in this [first tutorial](https://www.kernel-operations.io/keops/_auto_tutorials/a_LazyTensors/plot_lazytensors_b.html), this [2D example](https://www.kernel-operations.io/keops/_auto_tutorials/knn/plot_knn_torch.html) and this [K-NN classification script on MNIST](https://www.kernel-operations.io/keops/_auto_tutorials/knn/plot_knn_mnist.html). Just like the other KeOps operations, K-min reduction is both time- and memory-efficient: you never have to compute a full matrix of distances and can process large datasets on-the-fly. 
Is it suitable for you?

Best regards,
Jean

**Comment by srikanthmalla** [comment]
Hi @jeanfeydy ,

Got it! Thanks so much again :)

Best Regards,
Srikanth

**Comment by jeanfeydy** [comment]
Hi @srikanthmalla ,

You're very welcome: I have just upgraded the K-means tutorial to take your remarks into account 31ef415d9da1d77a19d9ce440d4ea68891bced60 . If you have any other question about the library / syntax, feel free to open a new issue :-) 

Best regards,
Jean

**Comment by srikanthmalla** [comment]
Hi @jeanfeydy ,
Do you also have any example for heirarchical clustering with your library? Please let me know.

Best Regards,
Srikanth
